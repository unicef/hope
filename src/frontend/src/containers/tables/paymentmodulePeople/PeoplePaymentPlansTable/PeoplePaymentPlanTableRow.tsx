import { Box, TableCell } from '@mui/material';
import * as React from 'react';
import { useNavigate } from 'react-router-dom';
import { useCashPlanVerificationStatusChoicesQuery } from '@generated/graphql';
import { BlackLink } from '@components/core/BlackLink';
import { StatusBox } from '@components/core/StatusBox';
import { ClickableTableRow } from '@components/core/Table/ClickableTableRow';
import { UniversalMoment } from '@components/core/UniversalMoment';
import {
  formatCurrencyWithSymbol,
  paymentPlanStatusToColor,
} from '@utils/utils';
import { useBaseUrl } from '@hooks/useBaseUrl';

interface PeoplePaymentPlanTableRowProps {
  plan;
  canViewDetails: boolean;
}

export const PeoplePaymentPlanTableRow = ({
  plan,
  canViewDetails,
}: PeoplePaymentPlanTableRowProps): React.ReactElement => {
  const navigate = useNavigate();
  const { baseUrl } = useBaseUrl();
  const paymentPlanPath = `/${baseUrl}/payment-module/${
    plan.isFollowUp ? 'followup-payment-plans' : 'payment-plans'
  }/${plan.id}`;
  const handleClick = (): void => {
    navigate(paymentPlanPath);
  };
  const { data: statusChoicesData } =
    useCashPlanVerificationStatusChoicesQuery();

  if (!statusChoicesData) return null;

  const followUpLinks = (): React.ReactElement => {
    if (!plan.followUps?.edges?.length) return <>-</>;
    return (
      <Box display="flex" flexDirection="column">
        {plan.followUps?.edges?.map((followUp) => {
          const followUpPaymentPlanPath = `/${baseUrl}/payment-module/followup-payment-plans/${followUp?.node?.id}`;
          return (
            <Box key={followUp?.node?.id} mb={1}>
              <BlackLink key={followUp?.node?.id} to={followUpPaymentPlanPath}>
                {followUp?.node?.unicefId}
              </BlackLink>
            </Box>
          );
        })}
      </Box>
    );
  };

  return (
    <ClickableTableRow
      hover
      onClick={canViewDetails ? handleClick : undefined}
      role="checkbox"
      key={plan.id}
    >
      <TableCell align="left">
        {plan.isFollowUp ? 'Follow-up: ' : ''}
        {canViewDetails ? (
          <BlackLink to={paymentPlanPath}>{plan.unicefId}</BlackLink>
        ) : (
          plan.unicefId
        )}
      </TableCell>
      <TableCell align="left">
        <StatusBox
          status={plan.status}
          statusToColor={paymentPlanStatusToColor}
        />
      </TableCell>
      <TableCell align="left">{plan.targetPopulation.name}</TableCell>
      <TableCell align="left">{plan.totalIndividualsCount || '-'}</TableCell>
      <TableCell align="left">{plan.currencyName}</TableCell>
      <TableCell align="right">
        {`${formatCurrencyWithSymbol(
          plan.totalEntitledQuantity,
          plan.currency,
        )}`}
      </TableCell>
      <TableCell align="right">
        {`${formatCurrencyWithSymbol(
          plan.totalDeliveredQuantity,
          plan.currency,
        )}`}
      </TableCell>
      <TableCell align="right">
        {`${formatCurrencyWithSymbol(
          plan.totalUndeliveredQuantity,
          plan.currency,
        )}`}
      </TableCell>
      <TableCell align="left">
        <UniversalMoment>{plan.dispersionStartDate}</UniversalMoment>
      </TableCell>
      <TableCell align="left">
        <UniversalMoment>{plan.dispersionEndDate}</UniversalMoment>
      </TableCell>
      <TableCell align="left">{followUpLinks()}</TableCell>
    </ClickableTableRow>
  );
};
