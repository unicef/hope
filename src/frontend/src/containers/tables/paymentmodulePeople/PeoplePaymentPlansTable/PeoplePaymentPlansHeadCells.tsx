export const headCells = [
  {
    disablePadding: false,
    label: 'Payment Plan ID',
    id: 'id',
    numeric: false,
  },
  {
    disablePadding: false,
    label: 'Status',
    id: 'status',
    numeric: false,
  },
  {
    disablePadding: false,
    label: 'Target Population',
    id: 'targetPopulation',
    numeric: false,
  },
  {
    disablePadding: false,
    label: 'Num. of People',
    id: 'totalIndividualsCount',
    numeric: false,
  },
  {
    disablePadding: false,
    label: 'Currency',
    id: 'currency',
    numeric: false,
  },
  {
    disablePadding: false,
    label: 'Total Entitled Quantity',
    id: 'totalEntitledQuantity',
    numeric: true,
  },
  {
    disablePadding: false,
    label: 'Total Delivered Quantity',
    id: 'totalDeliveredQuantity',
    numeric: true,
  },
  {
    disablePadding: false,
    label: 'Total Undelivered Quantity',
    id: 'totalUndeliveredQuantity',
    numeric: true,
  },
  {
    disablePadding: false,
    label: 'Dispersion Start Date',
    id: 'dispersionStartDate',
    numeric: false,
  },
  {
    disablePadding: false,
    label: 'Dispersion End Date',
    id: 'dispersionEndDate',
    numeric: false,
  },
  {
    disablePadding: false,
    label: 'Follow-up Payment Plans',
    id: 'followup-id',
    numeric: false,
  },
];
