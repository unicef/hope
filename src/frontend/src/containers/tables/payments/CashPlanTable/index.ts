import { CashPlanTable } from './CashPlanTable';

export { CashPlanTable };
