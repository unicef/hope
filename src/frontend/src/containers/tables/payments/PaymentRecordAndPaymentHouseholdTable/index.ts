import { PaymentRecordHouseholdTable } from './PaymentRecordAndPaymentHouseholdTable';

export { PaymentRecordHouseholdTable };
