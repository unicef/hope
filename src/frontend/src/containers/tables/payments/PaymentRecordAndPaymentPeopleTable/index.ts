import { PaymentRecordAndPaymentPeopleTable } from './PaymentRecordAndPaymentPeopleTable';

export { PaymentRecordAndPaymentPeopleTable };
