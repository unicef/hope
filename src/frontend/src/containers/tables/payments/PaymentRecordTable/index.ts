import { PaymentRecordTable } from './PaymentRecordTable';

export { PaymentRecordTable };
