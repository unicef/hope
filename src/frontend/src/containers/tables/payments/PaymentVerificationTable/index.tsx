import { PaymentVerificationTable } from './PaymentVerificationTable';

export { PaymentVerificationTable };
