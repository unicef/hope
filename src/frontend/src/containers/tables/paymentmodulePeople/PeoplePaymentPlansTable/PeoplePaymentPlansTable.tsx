import { ReactElement } from 'react';
import { useTranslation } from 'react-i18next';
import {
  AllPaymentPlansForTableQueryVariables,
  PaymentPlanNode,
  useAllPaymentPlansForTableQuery,
} from '@generated/graphql';
import { useBaseUrl } from '@hooks/useBaseUrl';
import { UniversalTable } from '../../UniversalTable';
import { PeoplePaymentPlanTableRow } from './PeoplePaymentPlanTableRow';
import { headCells } from './PeoplePaymentPlansHeadCells';

interface PeoplePaymentPlansTableProps {
  filter;
  canViewDetails: boolean;
}

export const PeoplePaymentPlansTable = ({
  filter,
  canViewDetails,
}: PeoplePaymentPlansTableProps): ReactElement => {
  const { t } = useTranslation();
  const { programId, businessArea } = useBaseUrl();
  const initialVariables: AllPaymentPlansForTableQueryVariables = {
    businessArea,
    search: filter.search,
    status: filter.status,
    totalEntitledQuantityFrom: filter.totalEntitledQuantityFrom || null,
    totalEntitledQuantityTo: filter.totalEntitledQuantityTo || null,
    dispersionStartDate: filter.dispersionStartDate || null,
    dispersionEndDate: filter.dispersionEndDate || null,
    isFollowUp: filter.isFollowUp ? true : null,
    program: programId,
  };

  return (
    <UniversalTable<PaymentPlanNode, AllPaymentPlansForTableQueryVariables>
      defaultOrderBy="-createdAt"
      title={t('Payment Plans')}
      headCells={headCells}
      query={useAllPaymentPlansForTableQuery}
      queriedObjectName="allPaymentPlans"
      initialVariables={initialVariables}
      renderRow={(row) => (
        <PeoplePaymentPlanTableRow
          key={row.id}
          plan={row}
          canViewDetails={canViewDetails}
        />
      )}
    />
  );
};
