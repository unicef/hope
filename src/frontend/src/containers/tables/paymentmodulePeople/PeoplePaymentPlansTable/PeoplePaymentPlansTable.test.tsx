import { MockedProvider } from '@apollo/react-testing';
import { act } from '@testing-library/react';
import wait from 'waait';
import { fakeApolloAllPaymentPlansForTable } from '../../../../../fixtures/payments/fakeApolloAllPaymentPlansForTable';
import { render } from '../../../../testUtils/testUtils';
import { PeoplePaymentPlansTable } from './PeoplePaymentPlansTable';

describe('containers/tables/payments/PeoplePaymentPlansTable', () => {
  it('should render with data', async () => {
    const { container } = render(
      <MockedProvider
        addTypename={false}
        mocks={fakeApolloAllPaymentPlansForTable}
      >
        <PeoplePaymentPlansTable
          canViewDetails={false}
          filter={{
            search: '',
            dispersionStartDate: undefined,
            dispersionEndDate: undefined,
            status: [],
            totalEntitledQuantityFrom: null,
            totalEntitledQuantityTo: null,
            isFollowUp: null,
          }}
        />
      </MockedProvider>,
    );
    await act(() => wait(0)); // wait for response

    expect(container).toMatchSnapshot();
  });

  it('should render loading', async () => {
    const { container } = render(
      <MockedProvider
        addTypename={false}
        mocks={fakeApolloAllPaymentPlansForTable}
      >
        <PeoplePaymentPlansTable
          canViewDetails={false}
          filter={{
            search: '',
            dispersionStartDate: undefined,
            dispersionEndDate: undefined,
            status: [],
            totalEntitledQuantityFrom: null,
            totalEntitledQuantityTo: null,
            isFollowUp: null,
          }}
        />
      </MockedProvider>,
    );
    await act(() => wait(0)); // wait for response

    expect(container).toMatchSnapshot();
  });
});
