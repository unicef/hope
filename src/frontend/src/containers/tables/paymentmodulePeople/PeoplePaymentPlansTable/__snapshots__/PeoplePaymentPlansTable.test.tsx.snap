// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/tables/payments/PeoplePaymentPlansTable should render loading 1`] = `
<div>
  <div>
    Unexpected error
  </div>
</div>
`;

exports[`containers/tables/payments/PeoplePaymentPlansTable should render with data 1`] = `
<div>
  <div>
    Unexpected error
  </div>
</div>
`;
